crates/dbs-miniball/docs/DeviceVirtualization.md

```diff
@@ -0,0 +1,307 @@
+        .map_err(Error::SetupInterruptController)
+    }
+```
+
+### 串行控制台
+为了使 VMM 能够使用串行控制台，除了 `dbs-legacy-devices` crate 中介绍的仿真部分之外，VMM 还需要执行以下操作：
+
+- 定义串行后端
+- 设置中断 IRQ
+- cmdline中添加console=ttyS0
+- 将串行端口添加到总线（PIO 或 MMIO）
+- 事件处理（可选）
+
+第 6 ～ 9 行，创建 serial console。第 13 ～ 15 行，设置 irq 为 COM1_IRQ = 4，console manager 的 irq 为 3 和 4。第 17 ～ 20 行，要通过serial类型的console进入虚拟机的console，需要在虚拟机的cmdline中添加console=ttyS0。第 24 ～ 27 行，创建 PioAddressRange 定义串行端口的地址和大小。第 29 ～ 33 行，注册 serial console。第 36 行，通过 evnet manager 监听事件处理。
+
+```rust
+pub const COM1_IRQ: u32 = 4;
+pub const COM1_PORT1: u16 = 0x3f8;
+
+fn add_serial_console(&mut self) -> Result<()> {
+    // Create the serial console.
+    let interrupt_evt = EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?;
+    let serial = Arc::new(Mutex::new(SerialDevice::new(
+        interrupt_evt.try_clone().map_err(Error::EventFd)?,
+    )));
+
+    // Register its interrupt fd with KVM. IRQ line 4 is typically used for serial port 1.
+    // See more IRQ assignments & info: https://tldp.org/HOWTO/Serial-HOWTO-8.html
+    self.vm
+        .register_irqfd(&interrupt_evt, COM1_IRQ)
+        .map_err(Error::IrqManager)?;
+
+    self.kernel_cfg
+        .cmdline
+        .insert_str("console=ttyS0")
+        .map_err(Error::Cmdline)?;
+
+    // COM1_PORT1 defines the base port address for the COM devices.
+    // Since every COM device has 8 data registers so we register the pio address range as size 0x8.
+    let resources = [Resource::PioAddressRange {
+        base: COM1_PORT1,
+        size: 0x8,
+    }];
+
+    self.device_mgr
+        .lock()
+        .unwrap()
+        .register_device_io(serial.clone(), &resources)
+        .map_err(Error::BusError)?;
+
+    // Hook it to event management.
+    self.event_mgr.add_subscriber(Box::new(serial));
+
+    Ok(())
+}
+```
+
+### 键盘 i8042
+I8042 数据端口（IO 端口 0x60）用于读取从 I8042 设备或 I8042 控制器本身接收的数据，并将数据写入 I8042 设备或 I8042 控制器本身。第 11 行，设置 irq 为 1，具有非常高的优先级。第 13 ～ 18 行，0x60 和 0x64 是 i8042 设备使用的 io 端口。将 pio 地址范围从 0x60 - 0x64 注册到 I8042_DATA_PORT 以供 i8042 使用。
+
+```rust
+pub const I8042_DATA_PORT: u16 = 0x60;
+
+fn add_i8042_device(&mut self) -> Result<()> {
+    let reset_evt =
+        EventFdTrigger::new(EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?);
+    let i8042_device = Arc::new(Mutex::new(I8042Device::new(
+        reset_evt.try_clone().map_err(Error::EventFd)?,
+        Arc::new(I8042DeviceMetrics::default()),
+    )));
+
+    self.vm.register_irqfd(&reset_evt, 1)?;
+
+    let resources = [Resource::PioAddressRange {
+        // 0x60 and 0x64 are the io ports that i8042 devices used.
+        // We register pio address range from 0x60 - 0x64 with base I8042_DATA_PORT for i8042 to use.
+        base: I8042_DATA_PORT,
+        size: 0x5,
+    }];
+
+    self.device_mgr
+        .lock()
+        .unwrap()
+        .register_device_io(i8042_device, &resources)
+        .map_err(Error::BusError)?;
+
+    Ok(())
+}
+```
+
+## virtio 设备
```

Comment:
vmm-reference has virtio-net, would you cover this part?

Reply:
Sorry, I haven't implemented virtio net in Miniball yet.